book/src/rfc/const_generic_fn.md

```diff
@@ -0,0 +1,224 @@
+## Summary
+This RFC is to support variable-length arguments and returned arrays in functions.
+
+## Motivation
+In the current implementation, array initialization requires explicitly stating all element values, which becomes impractical for large arrays. Furthermore, the lack of support for constant variables as array sizes limits the ability to write functions that generically handle arrays of varying sizes. This also necessitates hardcoding loop ranges, reducing flexibility.
+
+Allowing to define array sizes with constant variables would enhance function reusability and improve modularization.
+
+## Detailed design
+One of the useful features of const generic is to allow the declaration of default array or return array with const variable as size. Before making it possible to unlock these features, we need a way to propagate the actual values for the constant variables. This is crucial for doing sanity checks during circuit generation, such as checking the bounds of array access.
+
+To propagate the constant values, it needs to retain the structures for the variables from the bottom up. These retained structures would be also useful for builtin functions to do sanity checks on the inputs and outputs. For the support of const generic on types in the future, it would probably need to propagate the actual values in the same way.
+
+### Glossary
+- const: a keyword that can be used in different places. For example, when used in front of an argument name, in a function signature, it dictates that the function must be called with a value decided at compile time (for example, with a literal, or a global constant)"
+
+### Resolving const variables
+When the size of an array declaration is a literal number, the size can be resolved by calling the `size_of` on the array type. Then the `ExprKind::ArrayAccess` can check the bounds of the access by referencing the number returned from `size_of` call on an array type.
+
+When it is a constant variable as array size, we can't rely on the `size_of` to determine the size of the array directly through the type anymore, as the type checker can't resolve these values under its current design. For example, the type checker can't determine wether the length of the array matches with the return type:
```

Comment:
I feel like you're talking about initializing the array, but the example shows an issue of type checking. Shouldn't you write:

```diff
fn thing(const len: Field) -> [Field; len] {
-  return [0; 3];
+  return [0; len];
}
```

as example instead?